### ISSUE TITLE
Problem with publish duplication during close method

### ISSUE THREAD

**Issue by MaltsevDmytro** [micrometer-metrics/micrometer#3955]
**Describe the bug**
My custom class `MetricsPublisher` extends `StepMeterRegistry` from `micrometer-core`:
```java
public class MetricsPublisher extends StepMeterRegistry {

     @Override
     public void publish() {
         // Some publish logic
     }

     @Override
     @Predestroy
     public void close() {
         super.close();
     }
}
```
And unit test for `close()` method:
```java
@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
public class MetricsPublisherTest {

    @Autowired
    private final MetricsPublisher metricsPublisher;

    void closeTest() {
         metricsPublisher.close();
         
         verifyPublishedMetrics(); // verify that 5 expected metrics published during the closing
    }

}
``` 

But I failed test with **10** interactions instead of **5**, because of [this](https://github.com/micrometer-metrics/micrometer/blob/6fadd17b1bdac81fb31e7b5812745d89179ea91a/micrometer-core/src/main/java/io/micrometer/core/instrument/step/StepMeterRegistry.java#L146) and [this](https://github.com/micrometer-metrics/micrometer/blob/6fadd17b1bdac81fb31e7b5812745d89179ea91a/micrometer-core/src/main/java/io/micrometer/core/instrument/push/PushMeterRegistry.java#L135) "publish" methods in `close()` hierarchy. Is that expected behaviour when no publish was done before close due to library logic?
Then I decided to test close method and add `publish()` method call before `close()` in test:
```java
@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
public class MetricsPublisherTest {

    @Autowired
    private final MetricsPublisher metricsPublisher;
    void closeTest() {
         metricsPublisher.publish();

         metricsPublisher.close();
         
         verifyPublishedMetrics(); // now verify that 10 expected metrics published during the test
     }
}
```
Now I expected 10 published metric according to that [check](https://github.com/micrometer-metrics/micrometer/blob/6fadd17b1bdac81fb31e7b5812745d89179ea91a/micrometer-core/src/main/java/io/micrometer/core/instrument/step/StepMeterRegistry.java#L142), but I receive **15** now and after debugging realize  that It is no way to set [variable](https://github.com/micrometer-metrics/micrometer/blob/6fadd17b1bdac81fb31e7b5812745d89179ea91a/micrometer-core/src/main/java/io/micrometer/core/instrument/push/PushMeterRegistry.java#L44) that responsible for indication of publishing in current step from my extended class. 

!!! So it will always double push metrics even if publish was happend before.

**Environment**

 - Micrometer version: `1.11.1`
 - Micrometer registry: `library default`
 - OS: `macOS`
 - Java version: `Azul Zulu 17.0.6`

**To Reproduce**
How to reproduce the bug:

Take code snippet from description

**Expected behavior**

First version of test class in description should push metrics only once.

**Additional context**
None

**Comment by MaltsevDmytro** [comment]
UPD: Also about `isClosed()` check in this [method](https://github.com/micrometer-metrics/micrometer/blob/6fadd17b1bdac81fb31e7b5812745d89179ea91a/micrometer-core/src/main/java/io/micrometer/core/instrument/push/PushMeterRegistry.java#L134C1-L134C1), because first interaction with variable returned by `isClosed` method happened after `super.close()` [method](https://github.com/micrometer-metrics/micrometer/blob/6fadd17b1bdac81fb31e7b5812745d89179ea91a/micrometer-core/src/main/java/io/micrometer/core/instrument/step/StepMeterRegistry.java#L156) call:
```java
@Override
public void close() {
    stop();
    if (config.enabled() && !isClosed()) {
        publishSafely();
    }
    super.close();
}
```

**Comment by izeye** [comment]
I can see the extra `publish()` call with zero values. It seems to be confusing, but I'm not sure if it's okay as a step meter registry. Ideally, it would be better if it didn't happen.

See https://github.com/micrometer-metrics/micrometer/pull/4009

**Comment by marcingrzejszczak** [comment]
@izeye @shakuzen after the latest commits is this still a problem?

**Comment by izeye** [comment]
@marcingrzejszczak This seems to be a duplicate of https://github.com/micrometer-metrics/micrometer/issues/4357.

See https://github.com/micrometer-metrics/micrometer/commit/2735d262#diff-a21f40bfc61cbd4fbaf64ba4b263a9cfac19122d91e879418d835d33bf90e3d6R455-R470

**Comment by jonatan-ivanov** [comment]
Indeed, I think this should be fixed by https://github.com/micrometer-metrics/micrometer/issues/4357 and there were other fixes too like https://github.com/micrometer-metrics/micrometer/issues/1882 or https://github.com/micrometer-metrics/micrometer/issues/3872.

@MaltsevDmytro Could you please check if this is still an issue for your using the latest version?
